Support for IBM Power and fix proxy bug affecting cloud-native app-only monitoring

jsm84 opened the pull request:
# Description

EDIT: Closed in favor of #865 and #866 

This PR brings application-only monitoring support to OpenShift on IBM Power and Z.

Z-support is affected by a bug where `ruxitagentproc.conf` is missing from the oneagent install.zip for the s390 platform.

The motivation behind this is of joint concern to Dynatrace, IBM and Red Hat, to target a growing customer audience that is requesting Dynatrace Operator on the IBM Power and Z platforms.

## How can this be tested?
### New Build Scripts Provided
To get the operator code to build (with CGO enabled) for Power and Z, it was required to build locally against a RHEL 8 host. I would prefer a container/UBI-8 based build solution, but that was not possible due to the lack of a UBI-based go-toolset v1.17+ when I started this back in January. There is a new RHEL-specific container build script (a modified version of `./build/deploy_local.sh`) provided for this, as well as a script for building the bundle and index images, as well as the catalog source yaml file for deploying the index onto an OpenShift cluster, for installation using OperatorHub which is required for Red Hat certification of the operator.

### Requirements to Build and Test
To test this yourself, you'd need the following:
* A RHEL 8 VM running natively on Power or Z in order to build the operator (native container-based builds are now possible but not provided with this PR) with the following installed
   * Go 1.18 installed from go.dev
   * The following additional RHEL packages: `git gcc gcc-c++ make device-mapper-devel golang podman podman-docker gpgme-devel libassuan-devel`
   * Operator SDK v1.16.0
   * Operator Registry (`opm`) v1.19.5
   * The following go binaries:
      * sigs.k8s.io/controller-tools/cmd/controller-gen@v0.3.0
      * sigs.k8s.io/kustomize/kustomize/v3
   * Helm v3
* Power or Z-based OpenShift cluster, to be able to test app-only monitoring on the respective platform

### Steps to Build:

With the above installed and running on the local RHEL 8 VM, the `deploy_rhel.sh` script is called as follows:

```
$ GOARCH=<s390x|ppc64le> LOCALBUILD=1 \
 IMG=quay.io/orgname/dynatrace-operator TAG=v0.6.0-<arch> \
./hack/build/deploy_rhel.sh
```
You should now have the operator image built and pushed to the quay repository specified, with the tag that was specified.

Next, prepare the bundle, index and catalog source for deployment onto OpenShift with the `bundle_deploy_ocp.sh` script:

```
$ ./hack/build/bundle_deploy_ocp.sh 0.6.0-<arch> <quay-orgname>
```

The above will create and push a bundle image, followed by an index image, to the `<quay-orgname>` with the repository of `dynatrace-operator-bundle:v0.6.0-<arch>` and `operator-index:v<timestamp>-<build#>`, respectively. You will also see a catalog source yaml file named <quay-orgname>-operators.catalogsource.yaml`, with the respective bundle image and `<quay-orgname>` specified within the catalog source.

To deploy the operator bundle on an OpenShift cluster, first login to OpenShift using `oc login <api-url>` and simply apply the catalog source:

```
oc create -f <quay-orgname>-operators.catalogsource.yaml
```

### Steps to Test
In order to test, install the operator from the web console (I have a script, also) into the `openshift-operators` (All Operators) namespace.

For an application to monitor, I install the IBM OpenLiberty v0.8.1 operator from here (requires building the application container on the specific platform): https://github.com/OpenShift-Z/openliberty-operator-ocpz 

I create the required secret, containing the dynatrace PaaS and API Tokens/Keys into the `openshift-operators` namespace.

By default, that application runs in the `openliberty-demo` namespace. I label this namespace with the label `monitor: applicationMonitoring` and deploy the following CR into the `openshift-operators` namespace:

```
apiVersion: dynatrace.com/v1beta1
kind: DynaKube
metadata:
  name: dynakube
  namespace: openshift-operators
spec:
  apiUrl: https://<uuid>.live.dynatrace.com/api

  namespaceSelector:
    matchLabels:
      monitor: applicationMonitoring

  oneAgent:
      useCSIDriver: false
```

You should then see application metrics visible in your Dynatrace Dashboard, for the specified live environment `<uuid>`.

Please ping me over slack or via this PR for further questions and discussion.

## Checklist
- [x] Unit tests have been updated/added
- [ ] PR is labeled accordingly

0sewa0 commented:
pls fill out the description, also the first 2 commits are from the release-0.6 pls don't push does in this PR (rebase your changes on master)

thank you 😃

luhi-DT commented on config/samples/kustomization.yaml:
```diff
@@ -1,3 +1,3 @@
 resources:
-  - classicFullStack.yaml
+  - cloudNativeFullStack.yaml
```
that is unfortunatly not possible, since cloudNative is not running on OLM, because it needs the CSI Driver, which can not be applied because we create in through a manifest and OLM does not support daemonsets in manifests

jsm84 replied:
Ammended/Removed this change (now defaults to `applicationMonitoring.yaml` as intended)

luhi-DT commented on config/olm/openshift/bundle-0.2.1.Dockerfile:
```diff
@@ -18,6 +18,6 @@ LABEL operators.operatorframework.io.test.config.v1=tests/scorecard/
 COPY 0.2.1/manifests /manifests/
 COPY 0.2.1/metadata /metadata/
-LABEL com.redhat.openshift.versions="v4.5,v4.6,v4.7"
+LABEL com.redhat.openshift.versions="v4.5,v4.6"
```
Already released version, we just keep the files for history purposes

jsm84 replied:
Ammended/Removed this change

luhi-DT commented on config/olm/openshift/bundle-0.2.2.Dockerfile:
```diff
@@ -18,6 +18,6 @@ LABEL operators.operatorframework.io.test.config.v1=tests/scorecard/
 COPY 0.2.2/manifests /manifests/
 COPY 0.2.2/metadata /metadata/
-LABEL com.redhat.openshift.versions="v4.5,v4.6,v4.7"
+LABEL com.redhat.openshift.versions="v4.5,v4.6"
```
Already released version, we just keep the files for history purposes

jsm84 replied:
Ammended/Removed this change.

luhi-DT commented on src/arch/s390.go:
```diff
@@ -0,0 +1,7 @@
+
+package arch
+
+const Arch = ArchS390
```
can you use dtclient.ArchS390 here aswell?

jsm84 replied:
Ammended this change:

I actually needed to do the inverse: take the Power architectures and fix their references to the package local (relative to the `arch` pkg) `ArchPPC` and `ArchPPCLE` vars.

jsm84 commented:
Thanks to all for the code reviews. There's obviously alot of lint that needs to be cleaned up, still. I'll work on rebasing this and ammend/resubmit as needed.

jsm84 commented:
Rebased, linted and ammended as necessary. I could squash these commits further, if it helps. Just LMK. 

I need to test further and commit unit tests, pending completion of other work for Z/s390x.

meik99 commented on src/webhook/mutation/pod_mutator.go:
```diff
@@ -465,7 +472,8 @@ func (m *podMutator) getBasicData(pod *corev1.Pod) (
 	image string,
 ) {
-	flavor = kubeobjects.GetField(pod.Annotations, dtwebhook.AnnotationFlavor, arch.FlavorMultidistro)
+	// arch.Flavor defaults to the platform-specific value
```
```suggestion

	// arch.Flavor defaults to the platform-specific value
```

jsm84 replied:
Missed this in the first commit, as the msg was hidden.

Fixed in latest commit.

meik99 commented on src/webhook/mutation/pod_mutator.go:
```diff
@@ -794,3 +800,24 @@ func silentErrorResponse(podName string, err error) admission.Response {
+	var scs corev1.Secret
+	err := m.apiReader.Get(context.TODO(), client.ObjectKey{Name: dtwebhook.SecretConfigName, Namespace: nsName}, &scs)
+	if err != nil {
+		podLog.Error(err, "Failed to fetch config secret", "name", dtwebhook.SecretConfigName, "namespace", nsName)
```
```suggestion
		podLog.Error(err, "failed to fetch config secret", "name", dtwebhook.SecretConfigName, "namespace", nsName)
```

jsm84 replied:
Fixed all init capped `podLog.Error` strings in pod_mutator.go.

meik99 commented on src/webhook/mutation/pod_mutator.go:
```diff
@@ -794,3 +800,24 @@ func silentErrorResponse(podName string, err error) admission.Response {
 }
+
+// getSecretConfigData fetches the SecretConfig (initSecret.Config) from the pod namespace
+func (m *podMutator) getSecretConfigData(nsName string) (*standalone.SecretConfig, error) {
```
```suggestion
func (m *podMutator) getSecretConfigData(namespace string) (*standalone.SecretConfig, error) {
```

meik99 commented on src/webhook/mutation/pod_mutator.go:
```diff
@@ -794,3 +800,24 @@ func silentErrorResponse(podName string, err error) admission.Response {
+
+// getSecretConfigData fetches the SecretConfig (initSecret.Config) from the pod namespace
+func (m *podMutator) getSecretConfigData(nsName string) (*standalone.SecretConfig, error) {
+	var scs corev1.Secret
```
```suggestion
	var configSecret corev1.Secret
```

jsm84 replied:
Missed this in the first commit, as the msg was hidden.

Fixed in latest commit.

meik99 commented on src/webhook/mutation/pod_mutator.go:
```diff
@@ -794,3 +800,24 @@ func silentErrorResponse(podName string, err error) admission.Response {
+		return nil, err
+	}
+
+	scj := []byte(scs.Data[standalone.SecretConfigFieldName])
```
```suggestion
	secretData := []byte(scs.Data[standalone.SecretConfigFieldName])
```

jsm84 replied:
Missed this in the first commit, as the msg was hidden.

Fixed in latest commit.

meik99 commented on src/webhook/mutation/pod_mutator.go:
```diff
@@ -794,3 +800,24 @@ func silentErrorResponse(podName string, err error) admission.Response {
+
+	scj := []byte(scs.Data[standalone.SecretConfigFieldName])
+
+	var sc standalone.SecretConfig
```
```suggestion
	var config standalone.SecretConfig
```

jsm84 replied:
Missed this in the first commit, as the msg was hidden.

Fixed in latest commit.

meik99 commented on src/webhook/mutation/pod_mutator.go:
```diff
@@ -794,3 +800,24 @@ func silentErrorResponse(podName string, err error) admission.Response {
+	var sc standalone.SecretConfig
+	err = json.Unmarshal(scj, &sc)
+	if err != nil {
+		podLog.Error(err, "Failed to unmarshal field into type 'struct'", "FieldName", standalone.SecretConfigFieldName, "SecretName", dtwebhook.SecretConfigName)
```
```suggestion
		podLog.Error(err, "failed to unmarshal secret", "FieldName", standalone.SecretConfigFieldName, "SecretName", dtwebhook.SecretConfigName)
```

jsm84 replied:
Fixed init capped string + rephrased.

meik99 commented on src/webhook/mutation/pod_mutator.go:
```diff
@@ -725,18 +733,16 @@ func updateContainerOneAgent(c *corev1.Container, dk *dynatracev1beta1.DynaKube,
-			})
+		secretConfig, err := m.getSecretConfigData(nsName)
+		if err != nil {
+			podLog.Error(err, "Failed to inject proxy info into pod", "name", pod.Name, "generatedName", pod.GenerateName)
```
```suggestion
			podLog.Error(err, "failed to inject proxy info into pod", "name", pod.Name, "generatedName", pod.GenerateName)
```

jsm84 replied:
Fixed init capped strings in all `podLog.Error` calls.

meik99 commented on src/webhook/mutation/pod_mutator.go:
```diff
@@ -725,18 +733,16 @@ func updateContainerOneAgent(c *corev1.Container, dk *dynatracev1beta1.DynaKube,
-					},
-				},
-			})
+		secretConfig, err := m.getSecretConfigData(nsName)
```
Why does the data have to be fetched and passed to the env var instead of setting the secret as the source like before?

jsm84 replied:
@meik99 The previous code was assuming that the `proxy` field was an actual field name within the secret, but it was actually embedded within a serialized json string value of the `data.config` field. This was preventing the monitored pod from starting, as k8s would fail it with `CreateContainerError` as the field `config.proxy` does not exist within the secret.

I guess you can't use k8s downward API to reference field names nested within serialized/string values of another field. Thus, I had to fetch the secret and unmarshal the json string value of the `config` field so that I could reference the value of the `proxy` field contained within it.

If there's a better or more elegant way of doing this, I'm open to all suggestions/critique.

jsm84 replied:
@meik99 Also, because I needed a client, I had to add the receiver to the `*podMutator` all the way up the call chain.

meik99 commented on src/webhook/mutation/pod_mutator.go:
```diff
@@ -296,13 +296,13 @@ func addToInitContainers(pod *corev1.Pod, installContainer corev1.Container) {
 }
 
-func updateContainers(pod *corev1.Pod, injectionInfo *InjectionInfo, ic *corev1.Container, dk dynatracev1beta1.DynaKube, deploymentMetadata *deploymentmetadata.DeploymentMetadata) {
+func (m *podMutator) updateContainers(pod *corev1.Pod, injectionInfo *InjectionInfo, ic *corev1.Container, dk dynatracev1beta1.DynaKube, deploymentMetadata *deploymentmetadata.DeploymentMetadata, nsName string) {
```
Since another parameter has been added, and the signature was too long before already, please refactor your code such that parameter list does not have more than two parameters.

jsm84 replied:
@meik99 I've done a slight refactor which sort of proliferated throughout the pod_mutator code. I've added two new fields to the `podMutator` struct for the `currentPodNS` and also for  the current/active `dynaKube` instance related to the incoming pod request. Please review.

meik99 commented on src/webhook/mutation/pod_mutator.go:
```diff
@@ -687,7 +695,7 @@ func updateInstallContainerOneAgent(ic *corev1.Container, number int, name strin
 
 // updateContainerOA sets missing preload Variables
-func updateContainerOneAgent(c *corev1.Container, dk *dynatracev1beta1.DynaKube, pod *corev1.Pod, deploymentMetadata *deploymentmetadata.DeploymentMetadata) {
+func (m *podMutator) updateContainerOneAgent(c *corev1.Container, dk *dynatracev1beta1.DynaKube, pod *corev1.Pod, deploymentMetadata *deploymentmetadata.DeploymentMetadata, nsName string) {
```
Since another parameter has been added, and the signature was too long before already, please refactor your code such that parameter list does not have more than two parameters.

jsm84 replied:
I've done a slight refactor which sort of proliferated throughout the pod_mutator code. I've added two new fields to the `podMutator` struct for the `currentPodNS` and also for  the current/active `dynaKube` instance related to the incoming pod request. 

Doing this in turn shaved two args from the function declaration. Please review.

jsm84 commented on src/webhook/mutation/pod_mutator.go:
```diff
@@ -343,6 +343,13 @@ func decorateInstallContainerWithOneAgent(ic *corev1.Container, injectionInfo *I
 		)
+
+		// Setting the arch env var is not required for x86
+		if arch.Arch != "x86" {
```
Fixed hard coded string.

jsm84 commented on hack/build/bundle_deploy_ocp.sh:
```diff
@@ -0,0 +1,55 @@
+
+echo Created ./${NAMESPACE}-operators.catalogsource.yaml
+if [[ -n "$(oc whoami > /dev/null 2>&1 | grep admin)" ]]; then
+	oc apply -f ${NAMESPACE}-operators.catalogsource.yaml
```
Fixed missing quotes.

jsm84 commented on hack/build/bundle_deploy_ocp.sh:
```diff
@@ -0,0 +1,55 @@
+      interval: 5m
+EOF
+
+echo Created ./${NAMESPACE}-operators.catalogsource.yaml
```
Fixed missing quotes.

jsm84 commented on hack/build/bundle_deploy_ocp.sh:
```diff
@@ -0,0 +1,55 @@
+if [[ -n "$(pwd | grep config$)" ]] && [[ -n "$(dirname $pwd | grep dynatrace-operator$)" ]]; then
+	cd ..
+elif [[ -z "$(pwd | grep dynatrace-operator$)" ]] || [[ ! -f ./Makefile ]]; then
+	echo ERROR: Must run from the operator project root. Exiting...
```
Fixed missing quotes.

jsm84 commented on hack/build/bundle_deploy_ocp.sh:
```diff
@@ -0,0 +1,55 @@
+
+if [[ -z "$VERSION" ]]; then
+	echo ERROR: Not Enough Arguments.
+	echo Usage:
```
Fixed missing quotes.

meik99 commented on src/webhook/mutation/pod_mutator.go:
```diff
@@ -211,34 +213,44 @@ func (m *podMutator) Handle(ctx context.Context, req admission.Request) admissio
 	}
 
+	m.currentPodNS = ns.Name
+	defer func() {
+		m.currentPodNS = ""
+	}()
```
this defer function seems unnecessary

jsm84 replied:
@meik99 I was simply following the pattern from https://github.com/Dynatrace/dynatrace-operator/blob/master/src/webhook/mutation/pod_mutator.go#L200-L202

I agree that it's probably unnecessary if a new podMutator object is created (and likely garbage collected afterward) every time the `Handler()` function is called and the new struct instance of course has zero values for these fields.

It did make me stop to wonder why `m.currentPodName = ""` was being deferred in the original code (does this not get a new podMutator on every new `Handler()` call?), and I took the precaution against leaking old/outdated pods that were previously injected. It was just a safer bet to follow the same pattern, IMHO.

If you disagree, I'm happy to remove it, but we should probably also remove the above linked `defer func()` for nullifying `m.currentPodName` also.

Please advise.

meik99 replied:
please remove both defers, they are weird

jsm84 replied:
@meik99 @chrismuellner I've removed all deferred functions where podMutator fields are nullified.

meik99 commented on src/webhook/mutation/pod_mutator.go:
```diff
@@ -211,34 +213,44 @@ func (m *podMutator) Handle(ctx context.Context, req admission.Request) admissio
 
-	if dk.FeatureDisableMetadataEnrichment() {
+	m.dynaKube = dk
+	defer func() {
+		m.dynaKube = dynatracev1beta1.DynaKube{}
+	}()
```
this defer function seems unnecessary

jsm84 replied:
@meik99 @chrismuellner I've removed all deferred functions where podMutator fields are nullified.

jsm84 commented:
> @jsm84 we'd love to merge this if you have time slightly_smiling_face

@chrismuellner just waiting on a timely response from @meik99 and I'm happy to fix that as soon as I hear back.

0sewa0 commented:
@jsm84  why did you move the `e2e` folder ?

jsm84 commented:
> @jsm84 why did you move the `e2e` folder ?

Hi @0sewa0, there was a broken import that prevented DTO from compiling at one point, so I moved the e2e folder back into `./src` to get it to build. I can certainly move it back and fix the broken import.

jsm84 commented:
> > @jsm84 why did you move the `e2e` folder ?
> 
> Hi @0sewa0, there was a broken import that prevented DTO from compiling at one point, so I moved the e2e folder back into `./src` to get it to build. I can certainly move it back and fix the broken import.

I've moved the `e2e` pkg back into `testing/` and it still compiles with no issues.

That _should_ do it, but LMK if anything else sticks out.

jsm84 commented:
Removed support for Power big-endian mode. I added this at Michael V's suggestion, but it makes no sense to keep it, since OpenShift does not support power for BE, nor do I have a build host for that architecture.

jsm84 commented:
Squashed all commits & resolved all file conflicts against main branch. Ready for final review and merge.

jsm84 commented:
@chrismuellner I've just pushed fixes for the CI unit test failures, and helm linting failures. Can we force a retest?

chrismuellner commented:
@jsm84 Checks can be run as soon as the merge conflicts are resolved!

jsm84 commented:
> @jsm84 hi, so about the proxy secret reading problem: I fixed it in #839 and #841 TLDR: I just readded the proxy key to the secret so it can be mounted I did this as I don't like the idea to read a secret everytime we inject into a pod. (also this is waaaay simpler sweat_smile )

@0sewa0 That is perfect! I agree that fetching the secret every time a pod gets injected is undesirable behavior, but I was reluctant to change the schema of the secret myself. Now I can move forward with merging the new multi-arch code.

Let's keep this PR open, to hold the changes to the charts/manifests that will ultimately merge prior to official support for Power/Z. I will open a new PR to merge only the code changes for multi-arch.

jsm84 commented:
Closed in favor of #865 and #866
